File: src/test/java/rustleund/fightingfantasy/framework/closures/impl/AdjustScaleClosureTest.java
package rustleund.fightingfantasy.framework.closures.impl;

import kotlin.jvm.functions.Function1;
import org.easymock.EasyMockSupport;
import org.junit.jupiter.api.Test;
import org.w3c.dom.Element;
import rustleund.fightingfantasy.framework.base.GameState;
import rustleund.fightingfantasy.framework.base.PlayerState;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.ByteArrayInputStream;
import java.util.ArrayList;
import java.util.Objects;

import static org.easymock.EasyMock.expect;
import static org.junit.jupiter.api.Assertions.*;

public class AdjustScaleClosureTest extends EasyMockSupport {

	private final static DocumentBuilder documentBuilder;

	static {
		try {
			documentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
		} catch (ParserConfigurationException e) {
			throw new RuntimeException(e);
		}
	}

	@Test
	public void testExecute() {
		Element element = createMock(Element.class);

		//

		expect(element.getAttribute("type")).andStubReturn("time");
		expect(element.getAttribute("amount")).andStubReturn("2");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		AdjustScaleClosure testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		GameState gameState = new GameState();
		PlayerState playerState = newPlayer();
		gameState.setPlayerState(playerState);

		int oldAmount = playerState.getTime().getCurrentValue();

		assertTrue(testee.execute(gameState));

		assertEquals(oldAmount + 2, playerState.getTime().getCurrentValue());

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("time");
		expect(element.getAttribute("amount")).andStubReturn("4");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getTime().getCurrentValue();

		assertTrue(testee.execute(gameState));

		assertEquals(oldAmount + 4, playerState.getTime().getCurrentValue());

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("luck");
		expect(element.getAttribute("amount")).andStubReturn("-1");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getLuck().getCurrentValue();

		assertTrue(testee.execute(gameState));

		assertEquals(oldAmount - 1, playerState.getLuck().getCurrentValue());

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("time");
		expect(element.getAttribute("amount")).andStubReturn("-1");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(true);
		expect(element.getAttribute("promptOnFail")).andStubReturn("false");
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getTime().getCurrentValue();

		assertTrue(testee.execute(gameState));

		assertEquals(oldAmount, playerState.getTime().getCurrentValue());

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("time");
		expect(element.getAttribute("amount")).andStubReturn("-1");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(true);
		expect(element.getAttribute("promptOnFail")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getTime().getCurrentValue();

		assertFalse(testee.execute(gameState));

		assertEquals(oldAmount, playerState.getTime().getCurrentValue());
		assertEquals("You cannot perform this action", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("time");
		expect(element.getAttribute("amount")).andStubReturn("1");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(true);
		expect(element.getAttribute("promptOnFail")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getTime().getCurrentValue();

		assertTrue(testee.execute(gameState));

		assertEquals(oldAmount + 1, playerState.getTime().getCurrentValue());
		assertEquals("-", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("provisions");
		expect(element.getAttribute("amount")).andStubReturn("3");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(true);
		expect(element.getAttribute("promptOnFail")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsValue")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		assertTrue(testee.execute(gameState));

		assertEquals(3, playerState.getProvisions().getCurrentValue());
		assertEquals("-", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("provisions");
		expect(element.getAttribute("amount")).andStubReturn("13");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(true);
		expect(element.getAttribute("promptOnFail")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsValue")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		assertFalse(testee.execute(gameState));

		assertEquals(12, playerState.getProvisions().getCurrentValue());
		assertEquals("You cannot perform this action", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("provisions");
		expect(element.getAttribute("amount")).andStubReturn("13");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(true);
		expect(element.getAttribute("promptOnFail")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsValue")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(true);
		expect(element.getAttribute("adjustInitialValue")).andStubReturn("true");
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		assertFalse(testee.execute(gameState));

		assertEquals(12, playerState.getProvisions().getCurrentValue());
		assertEquals("You cannot perform this action", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("provisions");
		expect(element.getAttribute("amount")).andStubReturn("13");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsValue")).andStubReturn("true");
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(true);
		expect(element.getAttribute("adjustInitialValue")).andStubReturn("true");
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		assertTrue(testee.execute(gameState));

		assertEquals(12, playerState.getProvisions().getCurrentValue());
		assertEquals(13, Objects.requireNonNull(playerState.getProvisions().getUpperBound()).intValue());
		assertEquals("-", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("provisions");
		expect(element.getAttribute("amount")).andStubReturn("13");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(false);
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(true);
		expect(element.getAttribute("adjustInitialValue")).andStubReturn("true");
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		assertTrue(testee.execute(gameState));

		assertEquals(12, playerState.getProvisions().getCurrentValue());
		assertEquals(25, Objects.requireNonNull(playerState.getProvisions().getUpperBound()).intValue());
		assertEquals("-", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("provisions");
		expect(element.getAttribute("amount")).andStubReturn("-.375");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsPercent")).andStubReturn("true");
		expect(element.getAttribute("round")).andStubReturn("");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		assertTrue(testee.execute(gameState));

		assertEquals(8, playerState.getProvisions().getCurrentValue());
		assertEquals("-", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("provisions");
		expect(element.getAttribute("amount")).andStubReturn("-.375");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsPercent")).andStubReturn("true");
		expect(element.getAttribute("round")).andStubReturn("up");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		assertTrue(testee.execute(gameState));

		assertEquals(8, playerState.getProvisions().getCurrentValue());
		assertEquals("-", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("provisions");
		expect(element.getAttribute("amount")).andStubReturn("-.375");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsPercent")).andStubReturn("true");
		expect(element.getAttribute("round")).andStubReturn("down");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		assertTrue(testee.execute(gameState));

		assertEquals(7, playerState.getProvisions().getCurrentValue());
		assertEquals("-", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("time");
		expect(element.getAttribute("amount")).andStubReturn(".5");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsPercent")).andStubReturn("true");
		expect(element.getAttribute("round")).andStubReturn("down");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		oldAmount = playerState.getProvisions().getCurrentValue();
		assertEquals(12, oldAmount);

		try {
			testee.execute(gameState);
			fail("Should have thrown exception");
		} catch (IllegalArgumentException e) {
			// expected
		}

		assertEquals(12, playerState.getProvisions().getCurrentValue());
		assertEquals("-", gameState.getMessage());

		gameState.clearMessage();

		verifyAll();
		resetAll();

		//

		expect(element.getAttribute("type")).andStubReturn("figglebutt");
		expect(element.getAttribute("amount")).andStubReturn(".5");
		expect(element.hasAttribute("promptOnFail")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsValue")).andStubReturn(false);
		expect(element.hasAttribute("useAmountAsPercent")).andStubReturn(true);
		expect(element.getAttribute("useAmountAsPercent")).andStubReturn("true");
		expect(element.getAttribute("round")).andStubReturn("down");
		expect(element.hasAttribute("adjustInitialValue")).andStubReturn(false);
		expect(element.hasAttribute("rollDiceAmount")).andStubReturn(false);
		expect(element.hasAttribute("negate")).andStubReturn(false);

		replayAll();

		testee = new AdjustScaleClosure(element, GameState::getPlayerState);

		playerState = newPlayer();
		gameState.setPlayerState(playerState);

		assertFalse(testee.execute(gameState));

		verifyAll();
		resetAll();
	}

	@Test
	public void testFormula() {
		String testElementString = """
				<adjustScale type="stamina" amount="0" formula="AMT-(ceil(roll(1)/2)*2)" />
				""";
		Function1<Integer, Integer> diceRoller = createMock(Function1.class);
		GameState gameState = new GameState();
		PlayerState playerState = newPlayer();
		int startingStamina = playerState.getStamina().getCurrentValue();
		gameState.setPlayerState(playerState);

		AdjustScaleClosure testee = new AdjustScaleClosure(loadElement(testElementString), GameState::getPlayerState, diceRoller);

		expect(diceRoller.invoke(1)).andReturn(3);

		replayAll();

		assertTrue(testee.execute(gameState));
		assertEquals(startingStamina - 4, playerState.getStamina().getCurrentValue());

		verifyAll();
		resetAll();

		//

		playerState.getStamina().adjustCurrentValue(4);

		expect(diceRoller.invoke(1)).andReturn(2);

		replayAll();

		assertTrue(testee.execute(gameState));
		assertEquals(startingStamina - 2, playerState.getStamina().getCurrentValue());

		verifyAll();
		resetAll();
	}

	private PlayerState newPlayer() {
		return new PlayerState("test", new ArrayList<>());
	}

	private Element loadElement(String elementString) {
		try {
			return documentBuilder.parse(new ByteArrayInputStream(elementString.getBytes())).getDocumentElement();
		} catch (Exception e) {
			throw new RuntimeException(e);
		}
	}
}

File: src/test/kotlin/rustleund/fightingfantasy/framework/closures/impl/AdjustScaleClosureTest.kt
package rustleund.fightingfantasy.framework.closures.impl

import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withData
import rustleund.fightingfantasy.framework.base.GameState
import rustleund.fightingfantasy.framework.base.PlayerState
import rustleund.fightingfantasy.framework.base.Scale
import rustleund.test.TestElement
import strikt.api.expect
import strikt.assertions.*
import java.lang.IllegalArgumentException

class AdjustScaleClosureTest : FunSpec({
    context("Adjust Scales With Different Attributes") {
        withData(
            mapOf(
                "time + 2" to TestData(mapOf("type" to "time", "amount" to "2"), expectedValueAfterChange = 2),
                "time + 4" to TestData(mapOf("type" to "time", "amount" to "4"), expectedValueAfterChange = 4),
                "time - 1 no prompt" to TestData(mapOf("type" to "time", "amount" to "-1", "promptOnFail" to "false")),
                "time - 1 prompt fail" to TestData(
                    attributes = mapOf("type" to "time", "amount" to "-1", "promptOnFail" to "true"),
                    expectFailure = true,
                    expectedGameMessage = "You cannot perform this action"
                ),
                "time + 1 prompt fail" to TestData(
                    attributes = mapOf("type" to "time", "amount" to "1", "promptOnFail" to "true"),
                    expectedValueAfterChange = 1,
                ),
                "provisions = 3 prompt fail use amount" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "3",
                        "promptOnFail" to "true",
                        "useAmountAsValue" to "true"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 3,
                ),
                "provisions = 13 prompt fail use amount" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "13",
                        "promptOnFail" to "true",
                        "useAmountAsValue" to "true"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 12,
                    expectFailure = true,
                    expectedGameMessage = "You cannot perform this action"
                ),
                "provisions = 13 prompt fail use amount adjust init" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "13",
                        "promptOnFail" to "true",
                        "useAmountAsValue" to "true",
                        "adjustInitialValue" to "true"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 12,
                    expectFailure = true,
                    expectedGameMessage = "You cannot perform this action"
                ),
                "provisions = 13 no prompt fail use amount adjust init" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "13",
                        "useAmountAsValue" to "true",
                        "adjustInitialValue" to "true"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 12,
                    expectedUpperBoundAfterChange = 13,
                ),
                "provisions = 13 no prompt fail adjust init" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "13",
                        "adjustInitialValue" to "true"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 12,
                    expectedUpperBoundAfterChange = 25,
                ),
                "provisions = -0.375 no prompt fail amount as percent" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "-.375",
                        "useAmountAsPercent" to "true",
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 8,
                ),
                "provisions = -0.375 no prompt fail amount as percent round up" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "-.375",
                        "useAmountAsPercent" to "true",
                        "round" to "up"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 8,
                ),
                "provisions = -0.375 no prompt fail amount as percent round down" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "-.375",
                        "useAmountAsPercent" to "true",
                        "round" to "down"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 7,
                ),
                "time = 0.5 no prompt fail amount as percent round down" to TestData(
                    attributes = mapOf(
                        "type" to "time",
                        "amount" to ".5",
                        "useAmountAsPercent" to "true",
                        "round" to "down"
                    ),
                    expectIllegalArgument = true,
                ),
                "unknown type = 0.5 no prompt fail amount as percent round down" to TestData(
                    attributes = mapOf(
                        "type" to "figglebutt",
                        "amount" to ".5",
                        "useAmountAsPercent" to "true",
                        "round" to "down"
                    ),
                    expectFailure = true
                ),
                "provisions = 0 no prompt fail formula die roll 3" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "0",
                        "formula" to "AMT-(ceil(roll(1)/2)*2)"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 8,
                    diceRoller = { 3 }
                ),
                "provisions = 0 no prompt fail formula die roll 2" to TestData(
                    attributes = mapOf(
                        "type" to "provisions",
                        "amount" to "0",
                        "formula" to "AMT-(ceil(roll(1)/2)*2)"
                    ),
                    scaleSelector = { it.provisions },
                    expectedValueBeforeChange = 12,
                    expectedValueAfterChange = 10,
                    diceRoller = { 2 }
                ),
            )
        ) { testData ->
            val element = TestElement(testData.attributes)
            val testee = AdjustScaleClosure(element, { it.playerState }, testData.diceRoller)

            val gameState = GameState()
            val playerState = PlayerState("test", listOf())
            gameState.playerState = playerState

            val scale = testData.scaleSelector(playerState)

            expect {
                that(scale.currentValue).isEqualTo(testData.expectedValueBeforeChange)

                if (testData.expectIllegalArgument) {
                    catching { testee.execute(gameState) }.isFailure().isA<IllegalArgumentException>()
                } else {
                    that(testee.execute(gameState)).isNotEqualTo(testData.expectFailure)
                    that(scale.currentValue).isEqualTo(testData.expectedValueAfterChange)

                    if (testData.expectedUpperBoundAfterChange != null) {
                        that(scale.upperBound).isNotNull().isEqualTo(testData.expectedUpperBoundAfterChange)
                    }

                    if (testData.expectedGameMessage != null) {
                        that(gameState.message).isEqualTo(testData.expectedGameMessage)
                        gameState.clearMessage()
                    }
                }
            }
        }
    }
})

private data class TestData(
    val attributes: Map<String, String>,
    val scaleSelector: (PlayerState) -> Scale = { it.time },
    val expectedValueBeforeChange: Int = 0,
    val expectedValueAfterChange: Int = 0,
    val expectedUpperBoundAfterChange: Int? = null,
    val expectIllegalArgument: Boolean = false,
    val expectFailure: Boolean = false,
    val expectedGameMessage: String? = null,
    val diceRoller: (Int) -> Int = { 1 },
)